File: src/chapters/nonlinear-equations/content/code/newton-nd.cpp
#include <vector>
#include <iostream>
#include <functional>
#include <cmath>
using namespace std;
using Vec = vector<double>;
using Mat = vector<Vec>;

Vec solve(Mat A, Vec b) {                          // Gaussian elimination
    int n = b.size();
    for (int k = 0; k < n; ++k) {
        int p = k;
        for (int i = k + 1; i < n; ++i) if (fabs(A[i][k]) > fabs(A[p][k])) p = i;
        swap(A[k], A[p]); swap(b[k], b[p]);
        for (int i = k + 1; i < n; ++i) { double f = A[i][k] / A[k][k];
            for (int j = k; j < n; ++j) A[i][j] -= f * A[k][j]; b[i] -= f * b[k]; }
    }
    Vec x(n);
    for (int i = n - 1; i >= 0; --i) { double s = b[i];
        for (int j = i + 1; j < n; ++j) s -= A[i][j] * x[j]; x[i] = s / A[i][i]; }
    return x;
}

double normInf(const Vec& v) { double m = 0; for (double e : v) m = max(m, fabs(e)); return m; }

// Newton's method for F(x) = 0 with Jacobian J(x).
Vec newton_system(function<Vec(Vec)> F, function<Mat(Vec)> J, Vec x, double tol = 1e-12, int max_iter = 100) {
    for (int k = 0; k < max_iter; ++k) {
        Vec Fx = F(x);
        if (normInf(Fx) < tol) return x;
        Vec neg(Fx.size()); for (size_t i = 0; i < Fx.size(); ++i) neg[i] = -Fx[i];
        Vec dx = solve(J(x), neg);
        for (size_t i = 0; i < x.size(); ++i) x[i] += dx[i];
    }
    return x;
}

int main() {
    auto F = [](Vec v) -> Vec { return {v[0]*v[0] + v[1]*v[1] - 4, v[0]*v[1] - 1}; };
    auto J = [](Vec v) -> Mat { return {{2*v[0], 2*v[1]}, {v[1], v[0]}}; };
    Vec x = newton_system(F, J, {2.0, 0.5});
    cout.precision(8);
    cout << "x = " << x[0] << " " << x[1] << "\n";
}

File: src/chapters/nonlinear-equations/content/code/newton-nd.rs
// Newton's method for F(x) = 0 with Jacobian J(x).
fn solve(mut a: Vec<Vec<f64>>, mut b: Vec<f64>) -> Vec<f64> {
    let n = b.len();
    for k in 0..n {
        let mut p = k;
        for i in k + 1..n { if a[i][k].abs() > a[p][k].abs() { p = i; } }
        a.swap(k, p); b.swap(k, p);
        for i in k + 1..n {
            let f = a[i][k] / a[k][k];
            for j in k..n { a[i][j] -= f * a[k][j]; }
            b[i] -= f * b[k];
        }
    }
    let mut x = vec![0.0; n];
    for i in (0..n).rev() {
        let mut s = b[i];
        for j in i + 1..n { s -= a[i][j] * x[j]; }
        x[i] = s / a[i][i];
    }
    x
}
fn newton_system<F: Fn(&[f64]) -> Vec<f64>, J: Fn(&[f64]) -> Vec<Vec<f64>>>(
    f: F, j: J, x0: &[f64], tol: f64, max_iter: usize,
) -> Vec<f64> {
    let n = x0.len();
    let mut x = x0.to_vec();
    for _ in 0..max_iter {
        let fx = f(&x);
        if fx.iter().fold(0.0_f64, |m, &v| m.max(v.abs())) < tol { return x; }
        let dx = solve(j(&x), fx);
        for i in 0..n { x[i] -= dx[i]; }
    }
    x
}
fn main() {
    let f = |v: &[f64]| vec![v[0] * v[0] + v[1] * v[1] - 4.0, v[0] * v[1] - 1.0];
    let j = |v: &[f64]| vec![vec![2.0 * v[0], 2.0 * v[1]], vec![v[1], v[0]]];
    println!("{:?}", newton_system(f, j, &[2.0, 0.5], 1e-12, 100));
}
